### PR TITLE
[lld/mac] Allow -segprot having stricter initprot than maxprot on mac

### PR DESCRIPTION
**Pull request by nico**
...including for catalyst.

The usecase for this is to put certain security-critical variables into a special segment/section that's mapped as read-only most of the time, and that temporary gets remapped as writeable when these variables are written to be the program. This protects against them being written to by heap spraying attacks. This special section should be mapped as read-only at program start, so using

`-segprot MY_PROTECTED_MEMORY_THINGER rw r`

to mark that segment as rw maxprot and r initprot is exactly what we want.

lld has so far rejected mismatching initprot and maxprot.

ld64 doesn't reject this, but silently writes initprot into both fields (!) It looks like this might not be fully intentional, see https://crbug.com/41495919#comment5 and http://crbug.com/41495919#comment8.

In any case, when postprocessing ld64's output to have different values for initprot and maxprot, the dynamic loader seems to do the right thing (see also the previous two links).

The same technique also works on Windows, using both link.exe and lld-link.exe using `/SECTION:myprotsect,R`.

So, since this is useful, allow it when targeting macOS, and make it do what you'd expect.

Since loader support for this on iOS is less clear, keep disallowing it there for now.

See the PR for the program I used to check that this seems to work. (I only checked on arm64 macOS 14.5 so far; will run this on many more systems on bots once this is merged and rolled in.)

### CODE REVIEW

**Comment by nico** [comment]
To test that it works on macOS:

```
% % cat -n prot.cc
     1	#include <assert.h>
     2	
     3	#include <mach/mach_init.h>
     4	#include <mach/mach_vm.h>
     5	#include <sys/mman.h>
     6	
     7	#define DECLARE_PROTECTED_DATA constinit
     8	#define DEFINE_PROTECTED_DATA \
     9	  constinit __attribute__((section("PROTECTED_MEMORY, protected_memory")))
    10	
    11	extern char __start_protected_memory __asm(
    12	    "section$start$PROTECTED_MEMORY$protected_memory");
    13	extern char __stop_protected_memory __asm(
    14	    "section$end$PROTECTED_MEMORY$protected_memory");
    15	
    16	void CheckMemoryReadOnly(const void* ptr) {
    17	  mach_port_t object_name;
    18	  vm_region_basic_info_64 region_info;
    19	  mach_vm_size_t size = 1;
    20	  mach_msg_type_number_t count = VM_REGION_BASIC_INFO_COUNT_64;
    21	
    22	  kern_return_t kr = mach_vm_region(
    23	      mach_task_self(), reinterpret_cast<mach_vm_address_t*>(&ptr), &size,
    24	      VM_REGION_BASIC_INFO_64, reinterpret_cast<vm_region_info_t>(&region_info),
    25	      &count, &object_name);
    26	  assert(kr == KERN_SUCCESS &&
    27	         region_info.protection == VM_PROT_READ);
    28	}
    29	
    30	void CheckMemoryReadWrite(const void* ptr) {
    31	  mach_port_t object_name;
    32	  vm_region_basic_info_64 region_info;
    33	  mach_vm_size_t size = 1;
    34	  mach_msg_type_number_t count = VM_REGION_BASIC_INFO_COUNT_64;
    35	
    36	  kern_return_t kr = mach_vm_region(
    37	      mach_task_self(), reinterpret_cast<mach_vm_address_t*>(&ptr), &size,
    38	      VM_REGION_BASIC_INFO_64, reinterpret_cast<vm_region_info_t>(&region_info),
    39	      &count, &object_name);
    40	  assert(kr == KERN_SUCCESS &&
    41	         region_info.protection == (VM_PROT_READ | VM_PROT_WRITE));
    42	}
    43	
    44	// -segprot PROTECTED_MEMORY rw r
    45	
    46	DEFINE_PROTECTED_DATA int foo = 0;
    47	
    48	int main() {
    49	  int r;
    50	
    51	  //r = mprotect((void*)&foo, sizeof(foo), PROT_READ);
    52	  //assert(r == 0);
    53	
    54	  CheckMemoryReadOnly(&foo);
    55	
    56	
    57	  r = mprotect((void*)&foo, sizeof(foo), PROT_READ | PROT_WRITE);
    58	  assert(r == 0);
    59	
    60	  CheckMemoryReadWrite(&foo);
    61	
    62	
    63	  r = mprotect((void*)&foo, sizeof(foo), PROT_READ);
    64	  assert(r == 0);
    65	
    66	  CheckMemoryReadOnly(&foo);
    67	}
```

```
% out/gn/bin/clang prot.cc -std=c++20 -segprot PROTECTED_MEMORY rw r -isysroot $(xcrun -show-sdk-path) -fuse-ld=lld
```

```         
% ./a.out
```

Without lld, this fails to map the section as rw with mprotect:

```
% out/gn/bin/clang prot.cc -std=c++20 -segprot PROTECTED_MEMORY rw r -isysroot $(xcrun -show-sdk-path)             
% ./a.out                                                                                             
Assertion failed: (r == 0), function main, file prot.cc, line 58.
```

To check catalyst, following https://github.com/nico/hack/blob/main/notes/catalyst.md:

```
% out/gn/bin/clang prot.cc -std=c++20 -segprot PROTECTED_MEMORY rw r -isysroot $(xcrun -show-sdk-path) -fuse-ld=lld -target arm64-apple-ios13.1-macabi 
% codesign --sign - a.out
% ./a.out
```

File: lld/MachO/Driver.cpp [hunk 1]
```diff
@@ -1882,9 +1882,20 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
     StringRef segName = arg->getValue(0);
     uint32_t maxProt = parseProtection(arg->getValue(1));
     uint32_t initProt = parseProtection(arg->getValue(2));
-    if (maxProt != initProt && config->arch() != AK_i386)
-      error("invalid argument '" + arg->getAsString(args) +
-            "': max and init must be the same for non-i386 archs");
+
+    bool allowsDifferentInitAndMaxProt =
+        config->platform() == PLATFORM_MACOS ||
+        config->platform() == PLATFORM_MACCATALYST;
```
Comment (zmodem) [inline lld/MachO/Driver.cpp:1888]: Maybe there should be a comment here that ios is currently more of a "don't know" than a hard no?

**Review by zmodem — APPROVED** [review]
Sounds reasonable to me.

File: lld/MachO/Driver.cpp (hunk 1 above)
Comment (nico) [reply lld/MachO/Driver.cpp:1889]: Sure, added a comment. Thanks!

**Pull request merged**